blockchain/src/codec.rs

```diff
@@ -0,0 +1,388 @@
+    for proof in block.proofs.iter() {
+        match proof {
+            Proof::Transient => dst.write_u8(b"type", 0)?,
+            Proof::Committed(path) => {
+                dst.write_u8(b"type", 1)?;
+                path.encode(dst)?;
+            }
+        }
+    }
+    Ok(())
+}
+
+fn read_block_txs(src: &mut impl Reader) -> Result<Vec<BlockTx>, ReadError> {
+    let len = src.read_u32()? as usize;
+    const BLOCK_TX_MIN_LENGTH: usize = 8 + 8 + 8 + 8 + 32 + 32 + 1 + 11 * 32 + 8;
+    src.read_vec_with(len, BLOCK_TX_MIN_LENGTH, read_block_tx)
+}
+
+fn write_block_txs(block_txs: &[BlockTx], dst: &mut impl Writer) -> Result<(), WriteError> {
+    dst.write_u32(b"block_txs length", block_txs.len() as u32)?;
+    block_txs.iter().map(|e| write_block_tx(e, dst)).collect()
+}
+
+trait ReaderExt: Reader + Sized {
+    fn read_u8_vec(&mut self) -> Result<Vec<u8>, ReadError> {
+        let len = self.read_u32()? as usize;
+        self.read_vec(len)
+    }
+
+    fn read_signature(&mut self) -> Result<Signature, ReadError> {
+        let s = self.read_u8x32()?;
+        let r = self.read_u8x32()?;
+        Ok(Signature {
+            s: Scalar::from_bits(s),
+            R: CompressedRistretto(r),
+        })
+    }
+
+    fn read_shortid_vec(&mut self) -> Result<ShortIDVec, ReadError> {
+        self.read_u8_vec().map(ShortIDVec)
+    }
+
+    fn read_blockid(&mut self) -> Result<BlockID, ReadError> {
+        self.read_u8x32().map(BlockID)
+    }
+
+    fn read_hash(&mut self) -> Result<Hash, ReadError> {
+        self.read_u8x32().map(Hash)
+    }
+}
+
+trait WriterExt: Writer + Sized {
+    fn write_u8_vec(&mut self, label: &'static [u8], vec: &[u8]) -> Result<(), WriteError> {
+        self.write_u32(b"vec length", vec.len() as u32)?;
+        self.write(label, vec)?;
+        Ok(())
+    }
+
+    fn write_signature(&mut self, sig: &Signature) -> Result<(), WriteError> {
+        self.write(b"scalar", sig.s.as_bytes())?;
+        self.write(b"key", sig.R.as_bytes())?;
+        Ok(())
+    }
+
+    fn write_shortid_vec(
+        &mut self,
+        label: &'static [u8],
+        vec: &ShortIDVec,
+    ) -> Result<(), WriteError> {
+        self.write_u8_vec(label, vec.0.as_ref())
+    }
+
+    fn write_blockid(
+        &mut self,
+        label: &'static [u8],
+        block_id: &BlockID,
+    ) -> Result<(), WriteError> {
+        self.write(label, block_id.as_ref())
+    }
+
+    fn write_hash(&mut self, label: &'static [u8], hash: &Hash) -> Result<(), WriteError> {
+        self.write(label, hash.0.as_ref())
+    }
+}
+
+impl<R: Reader> ReaderExt for R {}
+impl<W: Writer> WriterExt for W {}
+
+impl Decodable for Message {
+    fn decode(src: &mut impl Reader) -> Result<Self, ReadError>
+    where
+        Self: Sized,
+    {
+        let message_type_byte = src.read_u8()?;
+        let message_type = MessageType::try_from(message_type_byte)?;
+        match message_type {
+            MessageType::Block => {
```

Comment:
it's going to be easier to review these pairs of read/write methods for each type, if they were grouped under `Message::(en|de)code_$type(&mut impl Reader)`: e.g. `decode_block`, `decode_getblock` etc.